### ISSUE TITLE
Add automatic bank transaction updates 

### ISSUE THREAD

**Issue by 13thJenius** [moneymanagerex/android-money-manager-ex#1144]
A simple api, like plaid.com should do it, but you may need to add an exception to the GPL license for the use of a proprietary library.

**Comment by MisterY** [comment]
Could you elaborate on the required functionality?
Maybe it's just me but I would not let another app read my bank account data, Open Source or not, for example.

**Comment by 13thJenius** [comment]
There are a variety of APIs that can allow you to pull just transaction
data. It's my understanding that its very secure.
I'm thinking that you could automatically record each transaction based on
that data, like mint or level money. Mint uses their own bank connections,
but level money used plaid.

**Comment by velmuruganc** [comment]
Even I am looking for this functionality. Please implement asap.

**Comment by MisterY** [comment]
@velmuruganc, I'd wholeheartedly recommend you learn some Android development, implement this, and submit a pull request. It will probably be faster than waiting for someone else to do it.

**Comment by MisterY** [comment]
@velmuruganc, thanks for the effort. The best way to go might be to pick a goal and then focus on getting it done. Could be a simple thing, like making the project to build, then adding some minor functionality.
At some point you'd feel comfortable adding more complex features. 
I always recommend to watch at least the first few minutes of [this video](https://www.youtube.com/watch?v=TK1mBqKvIyU). :)

**Comment by velmuruganc** [comment]
@MisterY , I have started learning the android dev. Let see how it is going. Also i have downloaded the source code and getting below error when i open in Android Studio 2.3.2. Please help me,

8:11 AM	Gradle sync failed: Could not find com.android.tools.build:gradle:3.0.0.
				Searched in the following locations:
				file:/C:/Program Files/Android/Android Studio/gradle/m2repository/com/android/tools/build/gradle/3.0.0/gradle-3.0.0.pom
				file:/C:/Program Files/Android/Android Studio/gradle/m2repository/com/android/tools/build/gradle/3.0.0/gradle-3.0.0.jar
				https://jcenter.bintray.com/com/android/tools/build/gradle/3.0.0/gradle-3.0.0.pom
				https://jcenter.bintray.com/com/android/tools/build/gradle/3.0.0/gradle-3.0.0.jar
				https://maven.fabric.io/public/com/android/tools/build/gradle/3.0.0/gradle-3.0.0.pom
				https://maven.fabric.io/public/com/android/tools/build/gradle/3.0.0/gradle-3.0.0.jar
				https://oss.sonatype.org/content/repositories/snapshots/com/android/tools/build/gradle/3.0.0/gradle-3.0.0.pom
				https://oss.sonatype.org/content/repositories/snapshots/com/android/tools/build/gradle/3.0.0/gradle-3.0.0.jar
				Required by:
				project :

**Comment by MisterY** [comment]
Hi! Welcome to development. :)
You will need the version 3 of Android Studio to compile the project. Studio 3 is in the stable channel now.

**Comment by velmuruganc** [comment]
@MisterY , I am getting following error with Android Studio 3,

Error:(361, 0) Could not get unknown property 'crashlyticsApisecret' for project ':app' of type org.gradle.api.Project.
<a href="openFile:E:\AndroidStudioProjects\android-money-manager-ex-master\app\build.gradle">Open File</a>

![image](https://user-images.githubusercontent.com/29807533/32695347-3282221c-c77f-11e7-8649-f47db697408f.png)

**Comment by MisterY** [comment]
Hi, sorry for the late reply. Unfortunately, that error does not tell me much. I am by no means an Android expert as I'm only using it when contributing to a few Open Source projects in my spare time. If there is anything I can do to help, let me know.

**Comment by velmuruganc** [comment]
Ok..., Do you have any idea about "crashlytics Api secret". Do we set this property somewhere in the project or Do i need to add any environment variables.

**Comment by MisterY** [comment]
You can try the suggestion from the comment:
```
// fabric.properties file needs to be in /app folder.
```
or somehow disable Fabric. I'd appreciate if you found how to avoid the compilation issues for Dev branch. GnuCash for Android also has Fabric but I had no issues with their build. It would be nice if someone worked it out.

**Comment by velmuruganc** [comment]
Hi MisterY,

Gr8... Now I am able to build after installing Fabric plugin thru the suggestion link you provided. Thanks for you help.

**Comment by velmuruganc** [comment]
@MisterY , I wrote the code to read the sms when it is received, but i need information on following to continue my analysis,

1. I want to read the last 2 transactions to check the notes.
2. I want to read the payee and their default category
3. How do i add transaction or edit transaction. 
4. How can i take care of the different language strings.

Do we have any functions for this. It will be really good if you have any docs on this and will help new learns like me.

**Comment by MisterY** [comment]
Hi, @velmuruganc, that's good progress. As you will see, the code is the best and the most up-to-date documentation. Just look for places where we read and display transactions, payees, create/edit transactions, and where we read the strings from on all the forms.
If you have any specific questions, let me know.

**Comment by MisterY** [comment]
Have you seen #185? If you only want to read SMS, this functionality should already be in place.
You can spend some time testing it and confirming if it still works instead of rewriting it. :)

**Comment by velmuruganc** [comment]
@MisterY , I went thru #185, this will show the screen and user has to make selection either to accept and cancel. But here, I want transaction to be saved automatically without manual intervention. May be, I have to keep the transaction status as none, so that user can review and make it reconcile or what ever he wants. As of now, I have added the below UI to enable this automation process,

![image](https://user-images.githubusercontent.com/29807533/32976218-d459277c-cc37-11e7-9286-2cc9871e1dc2.png)

**Comment by MisterY** [comment]
Very nice. You could also add some logic that asks for the correct permissions when the option is enabled and also when it is to be executed. I would not normally ask for SMS access permissions unless the user explicitly wants MMEX to read their SMS messages. Some users are very sensitive to permissions the apps ask for.

**Comment by velmuruganc** [comment]
@MisterY , Do you have any idea on this error. Could you please help me to get rid of this error.

![image](https://user-images.githubusercontent.com/29807533/32986593-da2753d0-ccfb-11e7-83dd-7d40baa5a28a.png)

**Comment by velmuruganc** [comment]
@MisterY , I replaces "this" with getActivity() and it is working fine now.

**Comment by MisterY** [comment]
Yes, it requires a context and Fragment does not provide one. It has to be the parent Activity or the app context, depending on what is the usage.

**Comment by velmuruganc** [comment]
@MisterY , could you please tell me how to build .apk for beta version? Just i want to install the beta in my smartphone for testing, so that my prod/main/working app will not get disturbed.

Also I am started collecting the different sms that i get everyday and looking for some solution to extract the data. I will let you know once i create some logic for this. It will be good if you provide some sample sms from your side, so that i know the bank sms format in other countries (other than india)

**Comment by MisterY** [comment]
Hi! You can try to build an apk choosing Beta as flavor. However, I'm not sure that would work. You can confirm if it does. It should be fairly simple to reinstall the stable version once you are done with testing, however.
As for the SMS messages - I do not have any as I have never used this feature and I'm not aware that any of my banks provide such service. Therefore I could never test anything that has to do with SMS processing. Sorry!

**Comment by velmuruganc** [comment]
@MisterY , I select beta as flavor, but I am getting below error. Also same with release flavor and only debug works fine

Error:Execution failed for task ':app:fabricGenerateResourcesRelease'.
> Crashlytics Developer Tools error.

**Comment by velmuruganc** [comment]
@13thJenius , do you have any sample sms that you get from banks? If so could you please share with me.

**Comment by velmuruganc** [comment]
@MisterY , I could able to get rid of the error by using API key in the meta data and I could able to create signed apk for beta and installed in my device successfully.

Could you please let me know how exactly you are reading data from database. Also let me know which file I can look at for these implementation. These info will help me to learn quickly

**Comment by velmuruganc** [comment]
@MisterY , I could able to extract the information (Ref. below image) from the sms based of the sample sms what I had. Also started exploring the method "saveData()". Let me spend some more time on this to find out the way to inject the extracted data in to db.

![mmex extracted](https://user-images.githubusercontent.com/29807533/33194810-9366bdae-d0f8-11e7-955c-f4928d7f8a19.png)

**Comment by velmuruganc** [comment]
@MisterY , i got stuck in converting the extracted amount to Money. Could you please let me know how i can convert to money.

**Comment by velmuruganc** [comment]
@MisterY , I could able to convert the amount to Money using MoneyFactory, but however my transaction are not getting saved with below code even saveTransaction() is returning True. Could you please help me to resolve

-------------------------------
@Inject
BriteDatabase database;

private EditTransactionCommonFunctions mCommon;
	
ITransactionEntity model = AccountTransaction.create();
mCommon = new EditTransactionCommonFunctions(null, model, database);
				
mCommon.transactionEntity.setAccountId(parseInt(fromAccountID));
mCommon.transactionEntity.setTransactionType(TransactionTypes.Withdrawal);
mCommon.transactionEntity.setAmount(MoneyFactory.fromString(transAmount));
mCommon.transactionEntity.setAccountToId(Constants.NOT_SET);
mCommon.transactionEntity.setAmountTo(MoneyFactory.fromDouble(0));
mCommon.transactionEntity.setPayeeId(parseInt(transPayee[0]));
mCommon.transactionEntity.setCategoryId(parseInt(transPayee[2]));
mCommon.transactionEntity.setSubcategoryId(parseInt(transPayee[3]));
mCommon.transactionEntity.setNotes(msgBody);
mCommon.transactionEntity.setDate(new MmxDate().toDate());

saveTransaction()
					
private boolean saveTransaction() {
        AccountTransactionRepository repo = new AccountTransactionRepository(mContext);

        if (!mCommon.transactionEntity.hasId()) {
            // insert
            mCommon.transactionEntity = repo.insert((AccountTransaction) mCommon.transactionEntity);

            if (!mCommon.transactionEntity.hasId()) {
                Toast.makeText(mContext, R.string.db_checking_insert_failed, Toast.LENGTH_SHORT).show();
                Timber.w("Insert new transaction failed!");
                return false;
            }
        }  else {
            // update
            boolean updated = repo.update((AccountTransaction) mCommon.transactionEntity);
            if (!updated) {
                Toast.makeText(mContext, R.string.db_checking_update_failed, Toast.LENGTH_SHORT).show();
                Timber.w("Update transaction failed!");
                return false;
            }
        }
        return true;
    }

**Comment by MisterY** [comment]
What seems to be the error you are getting?

**Comment by velmuruganc** [comment]
@MisterY, it didn't throw any error and saveTransaction() is returning true.

**Comment by MisterY** [comment]
Do you see the transaction record in the database, then?

**Comment by velmuruganc** [comment]
@MisterY , no I dont see in database and App as well.

**Comment by MisterY** [comment]
Did you debug through the code and see what is hapenning?

**Comment by velmuruganc** [comment]
@MisterY , yes i did thru debug and found that saveTransaction() is returning true, but i dont see any records in db or app.

**Comment by MisterY** [comment]
So, what you are saying is that saveTransaction() returns true even though it does not save the transaction. Can you investigate why?

**Comment by velmuruganc** [comment]
@MisterY , Finally I found the reason why it is not showing up in the app through saveTransaction returns true. I could see the row added in the db, then found that the issue is with the Status. Everything is started working fine after adding line "mCommon.transactionEntity.setStatus("");" in my code.

**Comment by MisterY** [comment]
Good find! It could be that the transactions are filtered by status.

**Comment by velmuruganc** [comment]
@MisterY , I have created new pull request, could you please let me know how to check-in to beta version.

**Comment by MisterY** [comment]
Hi, @velmuruganc, thanks for the contribution! I'll have a look into more details soon.
Beta is always built from the dev branch and all the pull requests should go there.
Master is the stable branch and is merged once the stable version is published.

**Comment by velmuruganc** [comment]
@MisterY , it looks like the pull request is created for merging master  from Dev, please let me know how to create pull request for Dev branch.

**Comment by velmuruganc** [comment]
@MisterY , I think, now i have created correct PR #1179 for merging my changes to Dev. Please have a look. Also please let me know how to close #1175 which was created for merging Dev to Master. I think, #1175 should be closed without merging to master.

**Comment by MisterY** [comment]
Coming out in the next beta. Thank you for the contribution!

**Comment by Houd1ny** [comment]
How this functionality is working? Because I turned this option on and nothing happened after input sms from bank

**Comment by Houd1ny** [comment]
Many thanks to the author of this pull request.  But actually this functionality will not work in general case. I looked the source code of this pull request. All regexps are hardcoded. This functionality could work in general case only if user could specify regexp for his bank sms.
@MisterY I want to contribute to this feature. What do you think will be the best approach to solve an issue with automated bank transactions? Using tasker or "Process bank sms"?

**Comment by MisterY** [comment]
Thanks for the offer! I would prefer to externalize this functionality since there will be many variables involved, just as you mention.
On the other hand there are scripting apps, like Tasker, which are well suited for this. All that needs to be done is complete the functionality, or confirm if it works. So far, none of the requestors ever bothered to try it out.
This would be a pretty flexible solution, unlike the current one, and automation scripts could be shared and modified as needed.
If you are willing, please go for it. I don't use this functionality and that's why it was never 100% completed in the past. I have tested that an intent can open the new transaction window and have the user confirm by saving explicitly.

**Comment by Houd1ny** [comment]
@MisterY Oh, I spend too match time manually entering SMS into the app. So it will save my time.
Actually, I have already tested "intent.action.INSERT" in Tasker described in "Support for Tasker" and it worked. I did not test all properties, but everything seems ok. I could share JavaScript scripts that parse SMS and call intend then write a tutorial. 
Intend will everytime open "New transaction window" and it is a little bit annoying. It will be better if a transaction was saved with status "None" and then, when a user has time, converted by him to "Reconciled" with small changes if needed(option "silent" for example). If a payee is not known it will be some default category(in this case some info from SMS could be copied to "Notes").
There is also the problem that payee in SMS does not always a have good string representation like shop112121213, or Uber BV when in-app it is "Uber", there also could be another name for the same app "uber".  So there should be some mapping from sms payee to app payee. That`s maybe the most tricky part. But it will make this feature much better. What do you think about this mapping? How could this be implemented in app?
Thank you for your time

**Comment by MisterY** [comment]
Hmmm. As there is currently no place in the database for this and it is a mobile-only feature, I guess we could store the mappings into another config file. 
Auto saving requires users to 
- create mappings (UI + storage)
- create a new Payee or find existing
- decide on the Category
- automatically creating transactions without user's knowledge
- require SMS permission

However, it is not to complex, I believe. There might be just that one setting missing in the intent parsing and some logic to process this intent, find the Payee / create new, find appropriate Category, and save the transaction.
A GUI somewhere in the settings is needed for the mapping and enabling/disabling the feature along with the SMS permission.

**Comment by velmuruganc** [comment]
Could you please provide me the sample sms which doesn't do the automatic bank transactions. As of now, I am doing following,

1. extract the bank account no.,
2. find the matched account in MMEX
3. Extract the payee,
4. find the matched payee in MMEX, if matched then get the last used Category & Sub  
5. Based on the account no, find out the currency and extract the amount. If currency is different from base currency, then do the conversion as well.
6. Do the automatic transaction if step 2 & 4 matched, else pop-up the New Transaction window.

Here what needed is user must enter the account no. XXXXXXXXX3245 in the Account No field from Edit Account screen. If this doesnt work, then provide me the sample sms then I can find the fix or what is missing.

However, i dont like the New Transaction screen pop up for this. I want to put these all info as a notification saying that "Automatic bank trransaction fail", so that user can click this notification to popup the New Transaction screen. Please refer the issue 1188 for the same. Currently I am working on this.

Even I tried creating new payee automatically, but as @Houd1ny  said always these strings are not good, that's why I have started poping up the New Transaction Screen.

 Also we can add way to map the account no when processing bank transactions.

**** Basically I am a tester, but I have learned Andriod development and did these all implementation as I needed this feature on priority to minimize the manual entry. However I dint had much sample SMS to make it perfectly. If someone find any better way, then please let me know what i am missing, so that i can correct my self. 

**** Once again thanks for @MisterY for showing path for my Android development.

**Comment by Houd1ny** [comment]
@velmuruganc Many thank to your work. But the problem of your PR is that you making an assumption on user bank language and sms format([example](https://github.com/moneymanagerex/android-money-manager-ex/blob/fcc4ea08cf28c94fd8e8c57a3727f9209d200ba1/app/src/main/java/com/money/manager/ex/notifications/SmsReceiverTransactions.java#L140)).  It is possible that there are another keywords and you can not know all of them in compile time.
Example my sms
"04.03.2018 18:24 INTERNET UBER BV, NL 43.69UAH CARD**6642 ZALISHOK 125243.54 UAH OVER 0.00  DOSTUPNO 125243.54UAH"
UAH == Ukrainian money
ZALISHOK == money left
DOSTUPNO = money available

Pls note that I leave in Ukraine and sms is basically in ukrainian language, but with Latin letters(it`s cheaper)

**Comment by velmuruganc** [comment]
@Houd1ny , You are correct. My regex will work only in English and I have covered all SMS format for most of the bank operates in India. Rest will have issue with this.

**Comment by Houd1ny** [comment]
@velmuruganc, @MisterY  So I proposing this approach
When option "Process bank sms" is turned on, in user accout profiels will appear fields for regexp for this account bank smses. They will be filed by default, but user could change them. So, in this case, this will be solved for all users. 
Another thing is about mapping. When this option is on we could add new option in Entities like "Bank payee" and the will be list of  mappings from "Bank payee" to "app payee". All new "Bank payee" will be added to list and mapped to default "app payee" and of course user could change them.
Guys what do you think?

**Comment by MisterY** [comment]
With so many different bank notifications, we will *never* cover everything. It only raises the code complexity trying to create all the mappings, regex, and other settings. This is why I'd prefer to use something like Tasker, or Open Source versions - LibreTasks, Easer, etc. to process the SMS. It seems more suitable for scripting.
I would prefer to keep MMEX limited to *only* the personal finance features. 
All that should be offered on MMEX side is a way to create transactions. However, this must be explicitly enabled by the user in the settings and also display the notification whenever it occurs, for transparency. No transactions should be created without the user knowing about it.
So, I would prefer if the API (intent) for creating a transaction would be simplified so that it is easy to use (if not already). The mappings for names and other details also should happen outside MMEX, in the layer that links SMS to MMEX. There is really no need for MMEX to have a permission to read someone's SMS messages. People have been already complaining about it. This is the way I'd go about it.
You are, however, free to implement it as you see fit, as long as it does not hinder other functionality. Note, however, that such code also needs to be maintained in the future.

**Comment by Houd1ny** [comment]
@MisterY I totally agree with you. But I`m concern that feature with Tasker is too "hacky" for many users and require another app. So there will be less then 10 users using it).  Actually, It will be a very cool feature. 
The problem for mapping outside the app is that I will have to enter app payee manually and not select it from list. 
You have a point. I will think about good implementation for scripting app. With good tutorial, it could be good. I will work on it.
Thank you

**Comment by MisterY** [comment]
You have a point. The counter-arguments would be:

- there are so many SMS implementations by different banks. Who will maintain all possible combinations once 5K users start using this functionality? Do we want to publish a new version every time one of zillion banks changes a parameter in their SMS message?
- parsing scripts can easily be shared via forum or elsewhere, outside MMEX
- there are open source equivalent of Tasker, so no cost involved. 
- it is a separate functionality, so having additional separate app is preferable. Android allows calling existing functionality via Intents, making this approach convenient.
- there could be better options in the future. Exposing an intent receiver on the side of MMEX, this should be future-proof, unlike coding a specific implementation inside the app.

However, I understand the convenience of a nicely wrapped feature inside the app. But I am concerned about the maintainability of such approach. Can anyone guarantee updating this in a timely manner and handling user support requests in regards to various banks and formats out there?

Don't take this as a critique. Just some counter-arguments for a fruitful discussion. I am not sure which is the way to go but let's put some arguments out there and see what makes the most sense.
Cheers!

**Comment by velmuruganc** [comment]
@MisterY , Is there any way to store the RegEx in the MMEX db, if so, then may be we can give UI for user to update these as per their requirement. Is this make sense?

**Comment by MisterY** [comment]
We could use a field in the settings table. But providing a RegEx interface to an average user... I always remember that saying "if you have a problem and it requires regex to solve, then you have two problems". :)

**Comment by MisterY** [comment]
Here's a temporary list of possible automation apps to consider:

- [Easer](https://github.com/renyuneyun/Easer)
- https://github.com/biotinker/LibreTasks
- https://bitbucket.org/fschrofner/swip/
- http://llamalab.com/automate/

**Comment by Houd1ny** [comment]
@MisterY It is critics. And critics are ok.
counter-arguments for this
"there are so many SMS implementations by different banks"
As I said before for each account there will be an option to specify regExp or something more ez.
So each user will write a filter for his bank and no development from team will be neeed.

**Comment by MisterY** [comment]
Right. But that is only the first step and covers just a part of the problem. You still need to figure out the Payee and the Category, as well as parse different number, date, and currency formats. These all require additional configuration. And then you end up creating a Tasker subsystem within MMEX.

**Comment by Houd1ny** [comment]
@MisterY Ok you are right
It will be regexp with match groups. A date could format could be configured, the problem with Payee and category fixed with mapped entities. But yeeeeeeeeee, it`s too much. 
It is better to share scripts for automation apps so users could configure them for their needs following tutorial on site. After that we could remove "Process bank" feature as there is no need in sms permission.

**Comment by MisterY** [comment]
You can decide which ways is more convenient but I believe scripts are more flexible in this respect. If I were to do this, I'd invest time in scripting a solution.
However, since @velmuruganc invested a significant effort into this, there is clearly less effort to add at this stage. This should also be intuitive enough if it is located in the app. Expecting users to know regex is a bit too much in my opinion. Anyone who is comfortable with regex will prefer a scripted solution because it is much more flexible. I mean, what if you have 2 banks? The hard-coded solution falls apart.
